.github/workflows/release.yml

```diff
@@ -0,0 +1,154 @@
+
+      - name: Set source date epoch
+        run: |
+          # Set SOURCE_DATE_EPOCH to the commit date of the last commit.
+          export SOURCE_DATE_EPOCH=$(git log -1 --pretty=%ct)
+          echo "SOURCE_DATE_EPOCH=$SOURCE_DATE_EPOCH" >> $GITHUB_ENV
+
+      - name: build
+        run: bin/rake build
+
+      - name: Check release and tag name match built version
+        run: |
+          for gem in pkg/*.gem; do
+            gemspec_version=$(gem spec ${gem} version | ruby -ryaml -e 'puts YAML.safe_load(ARGF.read, permitted_classes: [Gem::Version])')
+            if [ "${RELEASE_TAG_NAME}" != "v${gemspec_version}" ]; then
+              echo "Release tag name '${RELEASE_TAG_NAME}' does not match gemspec version 'v${gemspec_version}'"
+              exit 1
+            fi
+          done
+        env:
+          RELEASE_TAG_NAME: ${{ github.event.release.tag_name }}
+
+      - name: sign
+        run: |
+          # we smoke-test sigstore by installing each of the distributions
+          # we've built in a fresh environment and using each to sign and
+          # verify for itself, using the ambient OIDC identity
+          for dist in pkg/*; do
+            ./bin/smoketest "${dist}"
+          done
+
+      - name: Generate hashes for provenance
+        shell: bash
+        id: hash
+        working-directory: pkg
+        run: |
+          # sha256sum generates sha256 hash for all artifacts.
+          # base64 -w0 encodes to base64 and outputs on a single line.
+          # sha256sum artifact1 artifact2 ... | base64 -w0
+          echo "hashes=$(sha256sum * | base64 -w0)" >> $GITHUB_OUTPUT
+
+      - name: Save hashes
+        run: echo "$HASHES" | base64 -d > pkg/sha256sum.txt
+        env:
+          HASHES: ${{ steps.hash.outputs.hashes }}
+
+      - name: Upload built packages
+        uses: actions/upload-artifact@50769540e7f4bd5e21e526ee35c689e35e0d6874 # v4.4.0
+        with:
+          name: built-packages
+          path: ./pkg/
+          if-no-files-found: warn
+
+      - name: Upload smoketest-artifacts
+        uses: actions/upload-artifact@50769540e7f4bd5e21e526ee35c689e35e0d6874 # v4.4.0
+        with:
+          name: smoketest-artifacts
+          path: smoketest-artifacts/
+          if-no-files-found: warn
+
+  generate-provenance:
+    needs: [build]
+    name: Generate build provenance
+    permissions:
+      actions: read # To read the workflow path.
+      id-token: write # To sign the provenance.
+      contents: write # To add assets to a release.
+    # Currently this action needs to be referred by tag. More details at:
+    # https://github.com/slsa-framework/slsa-github-generator#verification-of-provenance
+    uses: slsa-framework/slsa-github-generator/.github/workflows/generator_generic_slsa3.yml@v2.0.0
+    with:
+      provenance-name: provenance-sigstore-${{ github.event.release.tag_name }}.intoto.jsonl
+      base64-subjects: "${{ needs.build.outputs.hashes }}"
+      upload-assets: true
```

Comment:
Nonblocker, but for viz: I think we're going to remove the corresponding step for this in sigstore-python soon, since I'd like to unify it with the attestation work we're building on PyPI. But that doesn't immediately affect you so probably nothing to do here.